keps/83-workload-preemption/README.md

```diff
@@ -0,0 +1,474 @@
+-->
+
+- `<package>`: `<date>` - `<test coverage>`
+
+#### Integration tests
+
+<!--
+Describe what tests will be added to ensure proper quality of the enhancement.
+
+After the implementation PR is merged, add the names of the tests here.
+-->
+
+### Graduation Criteria
+
+<!--
+
+Clearly define what it means for the feature to be implemented and
+considered stable.
+
+If the feature you are introducing has high complexity, consider adding graduation
+milestones with these graduation criteria:
+- [Maturity levels (`alpha`, `beta`, `stable`)][maturity-levels]
+- [Feature gate][feature gate] lifecycle
+- [Deprecation policy][deprecation-policy]
+
+[feature gate]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
+[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+-->
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Alternatives
+
+The following APIs were initially proposed to enhance the control over
+preemption, but they were left out of this KEP for lack of strong use cases.
+
+We might add them back in the future, based on feedback.
+
+### Inform how costly is to interrupt a Workload
+
+A workload might have a known cost of interruption that varies over time.
+For example:
+Early in its execution, the Workload hasn't made much progress, so it can be
+preempted. Later, the Worload is on the path of doing significant progress, so
+it's best not to disturb it. Lastly, the Workload is expected to have made
+some checkpoints, so it's ok to disturb it.
+
+This could be expressed with the following configuration:
+
+```yaml
+apiVersion: kueue.x-k8s.io/v1alpha2
+kind: Workload
+metadata:
+  name: my-workload
+spec:
+  preemption:
+    disruptionCostMilestones:
+    - seconds: 60
```

Comment:
Make preemption decisions through duration time seems hard to control, we don't know how fast or how slow the workload performs, there're many influencing factors, like infra runtime performance, execution-time spikes due to the biz logic problems.

Reply:
This was mostly motivated by https://github.com/kubernetes-sigs/kueue/issues/83#issuecomment-1224602577, but turns out nobody likes it :)
So I'm leaving it out of this KEP (it's already in the alternatives section)

Reply:
At the least, I like the idea about  preemption costs.

Reply:
I think we need a way to communicate the cost based on the actual execution, but this can be left as a future extension.